src/backend/access/aocs/aocsam_handler.c

```diff
@@ -1222,12 +1245,208 @@ aoco_relation_copy_for_cluster(Relation OldHeap, Relation NewHeap,
+
+	/* Need to open a new segment file. */
+	if (scan->cur_seg < 0)
+	{
+		if (open_next_scan_seg(scan) < 0)
+		{
+			elog(DEBUG2, "Failed to open a new segment.");
+			scan->cur_seg = -1;
+			return false;
+		}
+		scan->cur_seg_row = 0;
+	}
+
+	/* No more segments left. */
+	if (scan->cur_seg >= scan->total_seg)
+	{
+		elog(DEBUG2, "No more segments left.");
+		return false;
+	}
+
+	if (scan->total_row > targetrow)
+	{
+		elog(ERROR, "Analyze block inconsistency: total rows scanned %lu, target row %lu",
+			scan->total_row, targetrow);
+	}
+
+	/* Look up for the target row in a current segment. */
+	seginfo = scan->seginfo[scan->cur_seg];
+	rows_prev_segments = scan->total_row - scan->cur_seg_row;
+	if (targetrow <= rows_prev_segments + seginfo->total_tupcount)
+	{
+		return true;
+	}
+
+	/* Look up for the target row among other segments */
+	while (true)
+	{
+		scan->total_row += seginfo->total_tupcount - scan->cur_seg_row;
+		scan->cur_seg_row = 0;
+
+		close_cur_scan_seg(scan);
+		elog(DEBUG2, "Skipping current segment: segno %d", scan->cur_seg);
+		segno = open_next_scan_seg(scan);
+		if ((segno < 0) || (segno >= scan->total_seg))
+		{
+			elog(DEBUG3, "No more varblocks left in a current segment.");
+			return false;
+		}
+
+		/* Skip a segment if it doesn't contain the target row. */
+		seginfo = scan->seginfo[scan->cur_seg];
+		if (scan->total_row + seginfo->total_tupcount >= targetrow)
+		{
+			break;
+		}
+	}
+
+	return true;
+}
+
 static bool
 aoco_scan_analyze_next_block(TableScanDesc scan, BlockNumber blockno,
                                    BufferAccessStrategy bstrategy)
 {
+	AttrNumber proj;
+	int64 targetrow;
+	int64 varblockFirstRow;	/* var block row position with respect to previous segments */
+	int64 varblockRemainingRows;
+	AOCSFileSegInfo *seginfo;
+	DatumStreamRead *ds;
+	TupleTableSlot *slot;
+	bool found;
+	int segno;
+	int64 totalrows;
+	int64 totalRemainingRows;
+
 	AOCSScanDesc aoscan = (AOCSScanDesc) scan;
-	aoscan->targetTupleId = blockno;
+
+	/* Convert block number to a target row in an ordered sequence of appendonly segments. */
+	targetrow = blockno * aoscan->analyze_block_size;
+
+	/* This should never happen */
+	if (targetrow > INT64_MAX)
+	{
+		elog(ERROR, "Target row is out on range");
+	}
+
+	/* Initialize AOC scan. */
+	if (aoscan->columnScanInfo.relationTupleDesc == NULL)
+	{
+		aoscan->columnScanInfo.relationTupleDesc = RelationGetDescr(aoscan->rs_base.rs_rd);
+		PinTupleDesc(aoscan->columnScanInfo.relationTupleDesc);
+		aocs_initscan(aoscan);
+	}
+
+	/* Find a segment with a target row (skip other ones without scan and increment counters) */
+	if (!aoco_analyze_find_segment(aoscan, targetrow))
+	{
+		elog(DEBUG3, "No segment exists for the target row %lu", targetrow);
+		return false;
+	}
+
+	/* Calculate rows to scan in a current block. */
+	totalrows = 0;
+	for (segno = 0; segno < aoscan->total_seg; segno++)
+	{
+		totalrows += aoscan->seginfo[segno]->total_tupcount;
```

Comment:
We should use `GetAOCSSSegFilesTotals()` under snapshot for it instead.

Reply:
fixed in https://github.com/arenadata/gpdb/pull/135/commits/eff85e01962f23eeb74c96e167ac2018c40d663a